File: likho-app/src/lib/updater.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, Event } from "@tauri-apps/api/event";
import { Store } from "@tauri-apps/plugin-store";

// Update information from the backend
export interface UpdateInfo {
  version: string;
  current_version: string;
  notes?: string;
  pub_date?: string;
  mandatory: boolean;
}

// Update check response
export interface UpdateCheckResponse {
  available: boolean;
  info?: UpdateInfo;
  error?: string;
}

// Update state
export interface UpdaterState {
  checking: boolean;
  updateAvailable: UpdateInfo | null;
  downloadProgress: number;
  downloading: boolean;
  error: string | null;
}

// Store keys
const SETTINGS_STORE = "app-settings.json";
const LAST_CHECK_KEY = "last_update_check";
const SKIPPED_VERSION_KEY = "skipped_version";
const AUTO_CHECK_KEY = "auto_update_check";

// Event listeners
let unlistenProgress: (() => void) | null = null;
let unlistenFinished: (() => void) | null = null;

/**
 * Initialize the updater service
 */
export async function initUpdater(): Promise<void> {
  // Listen for download progress events
  unlistenProgress = await listen<number>(
    "update-download-progress",
    (event: Event<number>) => {
      console.log(`Download progress: ${event.payload}%`);
    }
  );

  // Listen for download finished events
  unlistenFinished = await listen(
    "update-download-finished",
    () => {
      console.log("Download finished");
    }
  );

  // Schedule periodic checks if enabled
  const autoCheck = await isAutoCheckEnabled();
  if (autoCheck) {
    await checkUpdateOnStartup();
  }
}

/**
 * Cleanup event listeners
 */
export function cleanupUpdater(): void {
  unlistenProgress?.();
  unlistenFinished?.();
}

/**
 * Check for updates
 */
export async function checkForUpdates(): Promise<UpdateCheckResponse> {
  try {
    const result = await invoke<UpdateCheckResponse>("check_update");
    await setLastCheckTime();
    return result;
  } catch (error) {
    console.error("Failed to check for updates:", error);
    return {
      available: false,
      error: error instanceof Error ? error.message : String(error),
    };
  }
}

/**
 * Check for updates on app startup
 * Respects the skipped version setting
 */
export async function checkUpdateOnStartup(): Promise<void> {
  // Don't check if we already checked recently (within 1 hour)
  const lastCheck = await getLastCheckTime();
  const oneHourAgo = Date.now() - 60 * 60 * 1000;
  
  if (lastCheck && lastCheck.getTime() > oneHourAgo) {
    console.log("Skipping update check - checked recently");
    return;
  }

  const result = await checkForUpdates();
  
  if (result.available && result.info) {
    const skippedVersion = await getSkippedVersion();
    
    if (result.info.version === skippedVersion) {
      console.log(`Skipping update to ${skippedVersion} - user chose to skip`);
      return;
    }
    
    // Emit an event that the UI can listen for
    window.dispatchEvent(
      new CustomEvent("update-available", { detail: result.info })
    );
  }
}

/**
 * Download and install update
 */
export async function installUpdate(): Promise<void> {
  try {
    await invoke("install_update");
    // App will restart automatically after installation
  } catch (error) {
    console.error("Failed to install update:", error);
    throw error;
  }
}

/**
 * Skip a specific version
 */
export async function skipVersion(version: string): Promise<void> {
  const store = await Store.load(SETTINGS_STORE);
  await store.set(SKIPPED_VERSION_KEY, version);
  await store.save();
}

/**
 * Get the skipped version
 */
export async function getSkippedVersion(): Promise<string | null> {
  const store = await Store.load(SETTINGS_STORE);
  return await store.get<string>(SKIPPED_VERSION_KEY) || null;
}

/**
 * Clear the skipped version
 */
export async function clearSkippedVersion(): Promise<void> {
  const store = await Store.load(SETTINGS_STORE);
  await store.delete(SKIPPED_VERSION_KEY);
  await store.save();
}

/**
 * Set last check time
 */
async function setLastCheckTime(): Promise<void> {
  const store = await Store.load(SETTINGS_STORE);
  await store.set(LAST_CHECK_KEY, new Date().toISOString());
  await store.save();
}

/**
 * Get last check time
 */
async function getLastCheckTime(): Promise<Date | null> {
  const store = await Store.load(SETTINGS_STORE);
  const lastCheck = await store.get<string>(LAST_CHECK_KEY);
  return lastCheck ? new Date(lastCheck) : null;
}

/**
 * Check if auto-check is enabled
 */
export async function isAutoCheckEnabled(): Promise<boolean> {
  const store = await Store.load(SETTINGS_STORE);
  const value = await store.get<boolean>(AUTO_CHECK_KEY);
  return value !== false; // Default to true
}

/**
 * Set auto-check enabled
 */
export async function setAutoCheckEnabled(enabled: boolean): Promise<void> {
  const store = await Store.load(SETTINGS_STORE);
  await store.set(AUTO_CHECK_KEY, enabled);
  await store.save();
}

/**
 * Get current app version
 */
export async function getAppVersion(): Promise<string> {
  return await invoke<string>("get_app_version");
}

/**
 * Get app name
 */
export async function getAppName(): Promise<string> {
  return await invoke<string>("get_app_name");
}

/**
 * Get platform info
 */
export async function getPlatformInfo(): Promise<{
  os: string;
  arch: string;
  family: string;
}> {
  return await invoke("get_platform_info");
}

/**
 * Compare two semantic versions
 * Returns: -1 if v1 < v2, 0 if equal, 1 if v1 > v2
 */
export function compareVersions(v1: string, v2: string): number {
  const parts1 = v1.split(".").map(Number);
  const parts2 = v2.split(".").map(Number);
  
  for (let i = 0; i < Math.max(parts1.length, parts2.length); i++) {
    const p1 = parts1[i] || 0;
    const p2 = parts2[i] || 0;
    
    if (p1 < p2) return -1;
    if (p1 > p2) return 1;
  }
  
  return 0;
}

/**
 * Check if version is newer than current
 */
export function isNewerVersion(current: string, candidate: string): boolean {
  return compareVersions(candidate, current) > 0;
}

// Default export
export default {
  initUpdater,
  cleanupUpdater,
  checkForUpdates,
  installUpdate,
  skipVersion,
  getSkippedVersion,
  clearSkippedVersion,
  isAutoCheckEnabled,
  setAutoCheckEnabled,
  getAppVersion,
  getAppName,
  getPlatformInfo,
  compareVersions,
  isNewerVersion,
};

File: likho-app/src-tauri/src/commands/updater.rs
use tauri::{command, AppHandle};
use crate::updater::{check_for_updates, UpdateCheckResult};
use crate::updater::UpdateInfo;
use serde::{Deserialize, Serialize};
use std::sync::atomic::{AtomicBool, Ordering};
use once_cell::sync::Lazy;

static UPDATE_AVAILABLE: Lazy<AtomicBool> = Lazy::new(|| AtomicBool::new(false));
static CURRENT_UPDATE_INFO: std::sync::Mutex<Option<UpdateInfo>> = std::sync::Mutex::new(None);
static DOWNLOAD_PROGRESS: std::sync::Mutex<f32> = std::sync::Mutex::new(0.0);

/// Response for update check command
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateCheckResponse {
    pub available: bool,
    pub info: Option<UpdateInfo>,
    pub error: Option<String>,
}

/// Update status response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateStatus {
    pub checking: bool,
    pub update_available: bool,
    pub downloading: bool,
    pub download_progress: f32,
    pub dismissed: bool,
}

/// Check for updates command - callable from frontend
#[command]
pub async fn check_update(app: AppHandle) -> Result<UpdateCheckResponse, String> {
    match check_for_updates(app).await {
        Ok(UpdateCheckResult::Available(info)) => {
            UPDATE_AVAILABLE.store(true, Ordering::SeqCst);
            if let Ok(mut guard) = CURRENT_UPDATE_INFO.lock() {
                *guard = Some(info.clone());
            }
            Ok(UpdateCheckResponse {
                available: true,
                info: Some(info),
                error: None,
            })
        }
        Ok(UpdateCheckResult::NotAvailable) => {
            UPDATE_AVAILABLE.store(false, Ordering::SeqCst);
            Ok(UpdateCheckResponse {
                available: false,
                info: None,
                error: None,
            })
        }
        Ok(UpdateCheckResult::Error(e)) => Ok(UpdateCheckResponse {
            available: false,
            info: None,
            error: Some(e),
        }),
        Err(e) => Err(e),
    }
}

/// Download and install update command
#[command]
pub async fn install_update(app: AppHandle) -> Result<(), String> {
    // Set downloading state
    if let Ok(mut progress) = DOWNLOAD_PROGRESS.lock() {
        *progress = 0.0;
    }
    
    crate::updater::download_and_install(app).await
}

/// Skip a specific version
#[command]
pub async fn skip_version_cmd(app: AppHandle, version: String) -> Result<(), String> {
    crate::updater::skip_version(&app, version)?;
    // Clear update available flag
    UPDATE_AVAILABLE.store(false, Ordering::SeqCst);
    Ok(())
}

/// Get current update status
#[command]
pub async fn get_update_status() -> Result<UpdateStatus, String> {
    let progress = DOWNLOAD_PROGRESS.lock()
        .map(|p| *p)
        .unwrap_or(0.0);
    
    Ok(UpdateStatus {
        checking: false, // Would track from state in production
        update_available: UPDATE_AVAILABLE.load(Ordering::SeqCst),
        downloading: progress > 0.0 && progress < 100.0,
        download_progress: progress,
        dismissed: false, // Would track from storage in production
    })
}

/// Get download progress
#[command]
pub async fn get_download_progress() -> Result<f32, String> {
    DOWNLOAD_PROGRESS.lock()
        .map(|p| *p)
        .map_err(|_| "Failed to get progress".to_string())
}

/// Get current update info
#[command]
pub async fn get_current_update_info() -> Result<Option<UpdateInfo>, String> {
    CURRENT_UPDATE_INFO.lock()
        .map(|info| info.clone())
        .map_err(|_| "Failed to get update info".to_string())
}

/// Dismiss current update notification
#[command]
pub async fn dismiss_update() -> Result<(), String> {
    // In production, this would store the dismissed state
    UPDATE_AVAILABLE.store(false, Ordering::SeqCst);
    Ok(())
}

/// Get current app version
#[command]
pub fn get_app_version(app: AppHandle) -> String {
    app.package_info().version.to_string()
}

/// Get app name
#[command]
pub fn get_app_name(app: AppHandle) -> String {
    app.package_info().name.to_string()
}

/// Get platform information
#[command]
pub fn get_platform_info() -> serde_json::Value {
    serde_json::json!({
        "os": std::env::consts::OS,
        "arch": std::env::consts::ARCH,
        "family": std::env::consts::FAMILY,
    })
}
